ert/data/record/_record.py

```diff
@@ -0,0 +1,224 @@
+import pathlib
+from abc import ABC
+from collections import deque
+from enum import Enum
+from typing import Dict, List, Mapping, MutableMapping, Optional, Tuple, Union, cast
+
+from beartype import beartype
+from beartype.roar import BeartypeException  # type: ignore
+from pydantic import PositiveInt
+from ert.serialization import get_serializer
+
+number = Union[int, float]
+numerical_record_data = Union[
+    List[number],
+    Dict[str, number],
+    Dict[int, number],
+]
+blob_record_data = bytes
+record_data = Union[numerical_record_data, blob_record_data]
+record_collection = Tuple["Record", ...]
+RecordIndex = Tuple[Union[int, str], ...]
+
+
+def _build_record_index(
+    data: numerical_record_data,
+) -> RecordIndex:
+    if isinstance(data, MutableMapping):
+        return tuple(data.keys())
+    else:
+        return tuple(range(len(data)))
+
+
+class RecordValidationError(Exception):
+    pass
+
+
+class RecordType(str, Enum):
+    LIST_FLOAT = "LIST_FLOAT"
+    MAPPING_INT_FLOAT = "MAPPING_INT_FLOAT"
+    MAPPING_STR_FLOAT = "MAPPING_STR_FLOAT"
+    BYTES = "BYTES"
+
+
+class Record(ABC):
+    @property
+    def data(self) -> record_data:
+        pass
+
+    @property
+    def record_type(self) -> RecordType:
+        pass
+
+
+class BlobRecord(Record):
+    def __init__(self, data: blob_record_data) -> None:
+        self._record_type = RecordType.BYTES
+
+        try:
+            self._data = self._validate_data(data)
+        except BeartypeException as e:
+            raise RecordValidationError(str(e))
+
+    @beartype
+    def _validate_data(self, data: blob_record_data) -> blob_record_data:
+        return data
+
+    @property
+    def data(self) -> blob_record_data:
+        return self._data
+
+    @property
+    def record_type(self) -> RecordType:
+        return RecordType.BYTES
+
+    def __eq__(self, o: object) -> bool:
+        if isinstance(o, type(self)):
+            return self.__dict__ == o.__dict__
+        return False
+
+
+class NumericalRecord(Record):
+    def __init__(
+        self, data: numerical_record_data, index: Optional[RecordIndex] = None
+    ) -> None:
+        if isinstance(data, (set, frozenset, deque, tuple)):
+            data = [val for _, val in enumerate(data)]
+
+        try:
+            self._validate_data(data)
+        except BeartypeException as e:
+            raise RecordValidationError(str(e))
+        self._data = data
+
+        if index is None:
+            index = _build_record_index(data)
+        self._validate_index(data, index)
```

Comment:
Do you need call this validation if the index is build in the line directly above?

Reply:
There's no validation in `_build_record_index`. Could you clarify?

Reply:
No, but there is a `_build_record_index` in `_validate_index`, used to validate an index. So if you called `build_record_index` already, there is not need to validate it again, I assume?

Reply:
The first `_build_record_index` is to build the index if there is none, the second `_build_record_index` is an implementation detail of the validation. 😂 at least that's how my head sees it. Anyway, I think of a way to change this such that I still get runtime validation of the `index`.

Reply:
Not very important, I was just  noticing it as a small inefficiency.